runtime/lua/vim/fs.lua

```diff
@@ -334,6 +327,46 @@ function M.find(names, opts)
   return matches
 end
 
+--- Find the first parent directory containing a specific "marker", relative to a buffer's
+--- directory.
+---
+--- Example:
+---
+--- ```lua
+--- -- Find the root of a Python project
+--- vim.fs.findroot(0, {'pyproject.toml', 'setup.py' })
+---
+--- -- Find the root of a git repository
+--- vim.fs.findroot(0, '.git')
+---
+--- -- Find the parent directory containing any file with a .csproj extension
+--- vim.fs.findroot(0, function(name, path)
+---   return name:match('%.csproj$') ~= nil
+--- end)
+--- ```
+---
+--- @param bufnr integer Buffer number, or 0 for current buffer.
+--- @param marker (string|string[]|fun(name: string, path: string): boolean) A marker, or list
+---               of markers, to search for. If a function, the function is called for each
+---               evaluated item and should return true if {name} and {path} are a match.
+--- @return string? # Directory containing one of the given markers, or nil if no directory was
+--- found.
+function M.findroot(bufnr, marker)
+  assert(bufnr, 'missing required argument: bufnr')
+  assert(marker, 'missing required argument: marker')
+  local paths = M.find(marker, {
+    upward = true,
+    path = vim.fs.dirname(vim.api.nvim_buf_get_name(bufnr)),
+    stop = vim.uv.os_homedir(),
+  })
+
+  if #paths == 0 then
+    return nil
+  end
+
+  return vim.fs.dirname(paths[1])
```

Comment:
Given how trivial this function is, is there any gain in having this wrapper around `fs.find`? It's not using anything that's visible only to the editor, and I thought we wanted to stop having this kind of easily-breakable LSP utilities.

Reply:
I have a function like this in my config and I suspect many others do as well, so imo it would be useful to have.

Reply:
> Given how trivial this function is, is there any gain in having this wrapper around `fs.find`? It's not using anything that's visible only to the editor, and I thought we wanted to stop having this kind of easily-breakable LSP utilities.

It's trivial if you know what it does (and should do), but the details (direction, termination, relative to an `args.buf` is not obvious to non-experts -- this is one of the many paper cuts that make people complain about LSP being hard to set up in Neovim.

Also, this is not particularly LSP specific; any plugin that tracks "projects" by directory structure and is triggered by an autocommand can profit from it.